File: Easy/Print matrix in diagonal pattern.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
/*You are required to complete this method */

class Solution
{
  public:
    vector<int> matrixDiagonally(vector<vector<int>>&mat)
    {
        //Your code here
        int n = mat.size();
        map<int, vector<int>> mappa;
        
        for (int i=0; i<n; i++)
        {
            for (int j=0; j<n; j++)
            {
                mappa[i+j].push_back(mat[i][j]);
            }
        }
        
        vector<int> v;
        
        int k=0;
        
        for (auto it : mappa)
        {
           if (k%2 == 0 && k > 0)
           {
               reverse(it.second.begin(), it.second.end());
           }
           
           for (auto it2 : it.second)
           {
               v.push_back(it2);
           }
           
           k++;
        }
        
        return v;
         
    }
};


//{ Driver Code Starts.
int main()
{
    int T;
    cin>>T;
  
    while(T--)
    {
        int n;
        int k=0;
        //cin>>k;
        cin>>n;
        vector<vector<int>>mat(n, vector<int>(n));
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                cin >> mat[i][j];
            }
        }
        Solution obj;
        vector<int>ans = obj.matrixDiagonally(mat);
        for(auto i: ans)cout << i << " ";
        cout << "\n";
        
       
    }
}
// } Driver Code Ends

File: Easy/Print matrix in diagonal pattern.py
# Your task is to complete this function
'''
    Create variables i=0, j=0 to store the current indices of row and column
    Run a loop from 0 to n*n, where n is side of the matrix.
    
    Use a flag isUp to decide whether the direction is upwards or downwards. 
    
    Set isUp = true initially the direction is going upward.
    
    If isUp = 1 then start printing elements by incrementing column index and decrementing the row index.
    
    Similarly if isUp = 0, then decrement the column index and increment the row index.
    
    Move to the next column or row (next starting row and column
    
    Do this till all the elements get traversed.
'''

class Solution:
    def matrixDiagonally(self, mat):
        # code here
        n = len(mat)
        
        ans  = [mat[0][0]]
        flag = True
        
        for i in range(1, (n * 2) - 1):
            for j in range(max(0, i + 1 - n), min(i + 1, n)):
                if flag:
                    ans.append(mat[j][i - j])
                else: 
                    ans.append(mat[i - j][j])
            
            flag = not flag
        
        return ans
        
        


#{ 
 # Driver Code Starts
# Driver Program
if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n = list(map(int, input().strip().split()))
        arr = list(map(int, input().strip().split()))
        matrix = [[0 for i in range(n[0])]for j in range(n[0])]
        k=0
        for i in range(n[0]):
            for j in range(n[0]):
                matrix[i][j] = arr[k]
                k+=1
        a = Solution().matrixDiagonally(matrix)
        for x in a:
            print(x,end=' ')
        print('')
# Contributed by: Harshit Sidhwa
# } Driver Code Ends
